File: app/src/main/java/com/yourname/turf/sensor/StepCounterManager.kt
package com.yourname.turf.sensor

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager

class StepCounterManager(context: Context) : SensorEventListener {

    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val stepSensor = sensorManager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER)
    
    val isSensorAvailable: Boolean = stepSensor != null

    private var stepBaseline: Int = -1
    private var onStepUpdate: ((steps: Int) -> Unit)? = null
    
    fun start(onStepUpdate: (steps: Int) -> Unit) {
        this.onStepUpdate = onStepUpdate
        stepBaseline = -1 // Reset baseline
        if (isSensorAvailable) {
            sensorManager.registerListener(this, stepSensor, SensorManager.SENSOR_DELAY_UI)
        }
    }

    fun stop() {
        if (isSensorAvailable) {
            sensorManager.unregisterListener(this)
        }
        onStepUpdate = null
    }

    override fun onSensorChanged(event: SensorEvent?) {
        if (event == null || event.sensor.type != Sensor.TYPE_STEP_COUNTER) return
        
        val totalSteps = event.values.getOrNull(0)?.toInt() ?: return
        if (stepBaseline == -1) {
            stepBaseline = totalSteps
        }
        
        val liveSteps = totalSteps - stepBaseline
        onStepUpdate?.invoke(liveSteps)
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // No-op
    }

    companion object {
        const val STRIDE_METRES = 0.762f

        /**
         * Helper to estimate steps from distance in kilometers
         */
        fun estimateSteps(distanceKm: Double): Int {
            val distanceMetres = distanceKm * 1000.0
            return (distanceMetres / STRIDE_METRES).toInt()
        }

        /**
         * Helper to calculate distance in kilometers from steps
         */
        fun calculateDistanceKm(steps: Int): Double {
            return (steps * STRIDE_METRES) / 1000.0
        }
    }
}

File: lib/sensors/step_counter_manager.dart
import 'dart:async';
import 'package:pedometer/pedometer.dart';

class StepCounterManager {
  static const double strideMetres = 0.762;

  bool _isSensorAvailable = true;
  int _stepBaseline = -1;
  StreamSubscription<StepCount>? _subscription;
  
  bool get isSensorAvailable => _isSensorAvailable;

  /// Start listening to physical step counter sensor
  void start(void Function(int steps) onStepUpdate, void Function(Object error) onError) {
    _stepBaseline = -1;
    _isSensorAvailable = true;

    _subscription = Pedometer.stepCountStream.listen(
      (StepCount event) {
        final int totalSteps = event.steps;
        if (_stepBaseline == -1) {
          _stepBaseline = totalSteps;
        }
        final int liveSteps = totalSteps - _stepBaseline;
        onStepUpdate(liveSteps);
      },
      onError: (error) {
        print("Step counter sensor error/unavailable: $error");
        _isSensorAvailable = false;
        onError(error);
      },
      cancelOnError: false,
    );
  }

  /// Stop listening to the sensor
  void stop() {
    _subscription?.cancel();
    _subscription = null;
  }

  /// Helper to estimate steps from distance in kilometers
  static int estimateSteps(double distanceKm) {
    final double distanceMetres = distanceKm * 1000.0;
    return (distanceMetres / strideMetres).toInt();
  }

  /// Helper to calculate distance in kilometers from steps
  static double calculateDistanceKm(int steps) {
    return (steps * strideMetres) / 1000.0;
  }
}
